### ISSUE TITLE
SVM as last layer of LSTM

### ISSUE THREAD

**Issue by hitherburrito** [keras-team/keras#12960]
I am trying to implement a program that classifies News into 4 categories. It is being implemented using an **LSTM Network**. 
I want to **replace the softmax classifier in the output layer with an SVM**.
Any idea how to could be done? I tried changing the loss function to 'hinge' , but that didn't work out. Please suggest some way.
 
**The model defined in the code is as follows-**

model = Sequential()
model.add(Embedding(n_most_common_words, emb_dim, input_length=X.shape[1]))
model.add(SpatialDropout1D(0.7))
model.add(LSTM(64, dropout=0.7, recurrent_dropout=0.7))
model.add(Dense(4, activation='softmax'))
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['acc'])
print(model.summary())
history = model.fit(X_train, y_train, epochs=epochs, batch_size=batch_size,validation_split=0.2,callbacks=[EarlyStopping(monitor='val_loss',patience=7, min_delta=0.0001)])
accr = model.evaluate(X_test,y_test)

**Comment by LHillmann** [comment]
Answered here: https://github.com/keras-team/keras/issues/2588 Please close if this answers your question. 

Best,
Leon

**Comment by ys145** [comment]
> Answered here: #2588 Please close if this answers your question.
> 
> Best,
> Leon

I checked out that answer and tried implementing the give solution, but it didn't work. Changing the loss to hinge and adding the regularisation term didn't work out for an LSTM network. So I asked for an alternative.

**Comment by gledsonmelotti** [comment]
How are you? I am trying to implement an SVM within of the keras cost function using sklearn.svm and sklearn.model_selection. However, I always get errors. I believe the problem is to convert the y_true and y_pred tensor into a numpy array to be used in sklearn.svm. Then I need to convert the predicted results to tensor to be used in the cost function of keras (categorical_hinge).

Can anybody help me?
`model_input = Input(shape = (img_width, img_height, channel_axis))`
`x = Convolution2D_bn(model_input, 32, 3, 3, strides=(2, 2), padding='valid')`
`x = Convolution2D_bn(x, 32, 3, 3, padding='valid')`
`...`
`x = GlobalAveragePooling2D()(x)`
`x = Dense(4096, kernel_regularizer=l2(1e-4), name='Dense_1')(x)`
`x = Activation('relu', name='relu1')(x)`
`x = Dropout(DROPOUT)(x)`
`x = Dense(4096, kernel_regularizer=l2(1e-4), name='Dense_2')(x)`
`x = Activation('relu', name='relu2')(x)`
`model_output = Dropout(DROPOUT)(x)`
`model = Model(model_input, model_output)`
`model.summary()`

`import tensorflow as tf`
`from keras import backend as K`
`from sklearn.svm import SVC`
`from sklearn.model_selection import GridSearchCV`
`from sklearn.preprocessing import StandardScaler`
`from keras.losses import categorical_hinge`
#
`def custom_loss_value(y_true, y_pred):`
   ` X = K.eval(y_pred)`
    `print(X)`
    `Y = np.ravel(K.eval(y_true))`
    `Predict = []`
    `Prob = []`
    `scaler = StandardScaler()`
    `X = scaler.fit_transform(X)`
    `param_grid = {'C': [0.1, 1, 8, 10], 'gamma': [0.001, 0.01, 0.1, 1]}`
    `SVM = GridSearchCV(SVC(kernel='rbf',probability=True), cv=3, param_grid=param_grid, scoring='auc', verbose=1)`
    `SVM.fit(X, Y)`
    `Final_Model = SVM.best_estimator_`
    `Predict = Final_Model.predict(X)`
    `Prob = Final_Model.predict_proba(X)`
    `return categorical_hinge(tf.convert_to_tensor(Y, dtype=tf.float32), tf.convert_to_tensor(Predict, dtype=tf.float32))`
#
`sgd = SGD(lr=0.001, decay=1e-6, momentum=0.9, nesterov=True)`
`model.compile(loss=custom_loss_value, optimizer=sgd, metrics=['accuracy'])`

**Comment by FelixLiman** [comment]
Hello, I am currently working on SVM as Last Layer of LSTM too, as far as I got, you need to change the Dense on 4 node from Softmax into Tanh or Linear (Mine personally works on Tanh), and it works pretty good.

https://machinelearningmastery.com/how-to-choose-loss-functions-when-training-deep-learning-neural-networks/
